elastic/kibana
# [Http] Replace `buildNr` with `buildSha` in static asset paths

## Summary

Follow up of [first CDN PR](https://github.com/elastic/kibana/pull/169408). Primary focus is replacing our build nr with SHA that allows cache busting and maintains anti-collision properties.

## How to test

Start Kibana as usual navigating around the app with the network tab open in your browser of choice. Keep an eye out for any asset loading errors. It's tricky to test every possible asset since there are many permutations, but generally navigating around Kibana should work exactly as it did before regarding loading bundles and assets.

## Notes
* did a high-level audit of usages of `buildNum` in `packages`, `src` and `x-pack` adding comments where appropriate.
* In non-distributable builds (like dev) static asset paths will be prefixed with `XXXXXXXXXXXX` instead of Node's `Number.MAX_SAFE_INTEGER`
* Added some validation to ensure the CDN url is of the expected form: nothing trailing the pathname

### Checklist

Delete any items that are not applicable to this PR.


### Risk Matrix

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| We break some first or third party dependencies on existing asset routes | Med | High | Attempting to mitgate by serving static assets from both old and new paths where paths have updated to include the build SHA. Additioanlly: it is very bad practice to rely on the values of the static paths, but someone might be |
| Cache-busting is more aggressive | High | Low | Unlikely to be a big problem, but we are now scoping more static assets to a SHA and so every new Kibana release will require clients to, for example, download fonts again. |


### For maintainers
